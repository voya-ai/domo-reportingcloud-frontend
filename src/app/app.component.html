<div class="container-fluid">
  <div class="row">
    <nav class="col-sm-3 col-md-2 d-none d-sm-block bg-light sidebar">
      <ul class="nav nav-pills flex-column">
        <li class="nav-item">
          <a routerLink="/"
              routerLinkActive="active"
              class="nav-link">View reports</a>
        </li>
        <li class="nav-item">
          <a routerLink="/report/add"
              routerLinkActive="active"
              class="nav-link">Add report</a>
        </li>
      </ul>
    </nav>
  </div>
</div>

<main class="col-sm-9 ml-sm-auto col-md-10 pt-3" role="main">
  <router-outlet></router-outlet>
</main>
