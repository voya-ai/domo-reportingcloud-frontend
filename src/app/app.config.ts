import { OpaqueToken } from "@angular/core";
import { environment } from 'environments/environment';

export let APP_CONFIG = new OpaqueToken("app.config");

export interface IAppConfig {
    apiEndpoint: string;
}

export const AppConfig: IAppConfig = {
    apiEndpoint: environment.apiUrl
};
