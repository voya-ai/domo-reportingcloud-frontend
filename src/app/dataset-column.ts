export interface DatasetColumn {
  name: string,
  checked: boolean
}
