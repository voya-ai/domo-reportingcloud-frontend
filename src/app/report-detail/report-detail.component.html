<div *ngIf="report$ | async as report">
  <div class="container">
    <div class="row">
      <div class="col">
        <h1>{{ report.name }}</h1>
      </div>
    </div>
    <div class="row">
      <div class="col-6">
        <h2>Generate</h2>
        <div>
          <mat-select placeholder="Timespan" [(ngModel)]="timespan" name="timespan" class="form-control">
            <mat-option *ngFor="let range of ranges" [value]="range.value">
              {{ range.label }}
            </mat-option>
          </mat-select>
        </div>

        <div *ngIf="timespan != 0">
          <div>
            <mat-form-field>
              <input matInput [matDatepicker]="picker" placeholder="Choose a start date" [(ngModel)]="startDate">
              <mat-datepicker-toggle matSuffix [for]="picker"></mat-datepicker-toggle>
              <mat-datepicker #picker></mat-datepicker>
            </mat-form-field>
          </div>
          <div>
            <mat-select placeholder="Time column" [(ngModel)]="timeColumn" name="timeColumn" class="form-control">
              <mat-option *ngFor="let column of report.headerColumns" [value]="column">
                {{ column }}
              </mat-option>
            </mat-select>
          </div>
        </div>

        <div>
          <div class="form-group">
            <label>Master group (different folders)</label>
            <mat-select placeholder="Master group" [(ngModel)]="masterGroup" name="masterGroup" class="form-control">
              <mat-option>None</mat-option>
              <mat-option *ngFor="let column of report.headerColumns" [value]="column">
                {{ column }}
              </mat-option>
            </mat-select>
          </div>
        </div>

        <div>
          <div class="form-group">
            <label>Child group (different documents)</label>
            <mat-checkbox *ngFor="let column of childGroupColumns" [(ngModel)]="column.childGroup" class="form-check">
              {{column.name}}
            </mat-checkbox>
          </div>
        </div>

        <div>
          <div *ngIf="!loading; else showSpinner">
            <button
              type="button"
              class="btn btn-primary"
              (click)="generate(report)">
              Generate
            </button>
          </div>
          <ng-template #showSpinner>
            <mat-spinner></mat-spinner>
          </ng-template>
        </div>
      </div>
      <div class="col-6">
        <app-report-detail-sample
          [report]="report">
        </app-report-detail-sample>
      </div>
    </div>
    <div class="row">
      <div class="col">
        <app-report-edit
          [report]="report"
          (update)="update($event)">
        </app-report-edit>
      </div>
    </div>
  </div>
</div>
