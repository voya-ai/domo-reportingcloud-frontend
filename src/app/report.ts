export interface Report {
  name?: string;
  slug: string;
  dataSet?: string;
  templateName?: string;
  headerColumns?: string[];
  contentColumns?: string[];
}
