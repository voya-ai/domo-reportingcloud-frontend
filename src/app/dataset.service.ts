import { Injectable, Inject } from '@angular/core';
import { Http } from '@angular/http';
import { Observable } from 'rxjs/Observable';
import { APP_CONFIG, IAppConfig } from './app.config';
import { DatasetColumn } from './dataset-column';

import 'rxjs/add/operator/map';
import 'rxjs/add/operator/filter';
import 'rxjs/add/operator/first';
import 'rxjs/add/operator/shareReplay';

@Injectable()
export class DatasetService {
  private data : Observable<any[]>;

  constructor(
    private _http: Http,
    @Inject(APP_CONFIG) private config
  ) {
    this.data = this._http.get(`${this.config.apiEndpoint}/datasets`)
      .map(res => res.json())
      .shareReplay(1);
  }

  search(term: string) : Observable<any[]> {
    return this.data
      .map
      (data => {
        return data.filter(item => {
          return item.name.search(new RegExp(term, "i")) !== -1;
        });
      });
  }

  get(id: string) : Observable<any> {
    return this._http.get(`${this.config.apiEndpoint}/datasets/${id}`)
      .map(res => res.json());
  }

  getColumns(id: string) : Observable<DatasetColumn[]> {
    return this.get(id)
      .map(res => res.schema.columns.map(column => {
        return {
          name: column.name,
          headerChecked: false,
          contentChecked: false
        }
      })).map(data => {
        return data.sort((a, b) => {
          if(a.name < b.name) return -1;
          if(a.name > b.name) return 1;
          return 0;
        });
      });
  }

}
