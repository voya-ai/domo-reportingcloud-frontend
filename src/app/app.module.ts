import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { HttpModule } from '@angular/http';

import { RouterModule, Routes } from '@angular/router';

import {
  MatAutocompleteModule,
  MatInputModule,
  MatCardModule,
  MatGridListModule,
  MatCheckboxModule,
  MatButtonModule,
  MatDatepickerModule,
  MatNativeDateModule,
  MatSelectModule,
  MatMenuModule,
  MatListModule,
  MatToolbarModule,
  MatProgressSpinnerModule
} from '@angular/material';

import { AppComponent } from './app.component';

import { APP_CONFIG, AppConfig } from './app.config';

import { ReportService } from './report.service';
import { DatasetService } from './dataset.service';
import { TemplateService } from './template.service';
import { ReportGenerationService } from './report-generation.service';

import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { ReportListComponent } from './report-list/report-list.component';
import { ReportAddComponent } from './report-add/report-add.component';
import { ReportDetailComponent } from './report-detail/report-detail.component';
import { ReportDatasetAutocompleteComponent } from './report-dataset-autocomplete/report-dataset-autocomplete.component';
import { ReportTemplateAutocompleteComponent } from './report-template-autocomplete/report-template-autocomplete.component';
import { ReportGenerationListComponent } from './report-generation-list/report-generation-list.component';
import { ReportsComponent } from './reports/reports.component';
import { ReportDetailSampleComponent } from './report-detail-sample/report-detail-sample.component';
import { ReportEditComponent } from './report-edit/report-edit.component';
import { DatasetColumnSelectComponent } from './dataset-column-select/dataset-column-select.component';

const appRoutes: Routes = [
    {
        path: 'report/add',
        component: ReportAddComponent
    },
    {
        path: 'report/:slug',
        component: ReportDetailComponent
    },
    {
        path: '',
        component: ReportsComponent,
        pathMatch: 'full'
    }
];

@NgModule({
  declarations: [
    AppComponent,
    ReportListComponent,
    ReportAddComponent,
    ReportDetailComponent,
    ReportDatasetAutocompleteComponent,
    ReportTemplateAutocompleteComponent,
    ReportGenerationListComponent,
    ReportsComponent,
    ReportDetailSampleComponent,
    ReportEditComponent,
    DatasetColumnSelectComponent
  ],
  imports: [
    BrowserModule,
    FormsModule,
    HttpModule,
    ReactiveFormsModule,
    NoopAnimationsModule,
    RouterModule.forRoot(appRoutes),
    MatAutocompleteModule,
    MatInputModule,
    MatCardModule,
    MatGridListModule,
    MatCheckboxModule,
    MatButtonModule,
    MatDatepickerModule,
    MatNativeDateModule,
    MatSelectModule,
    MatMenuModule,
    MatListModule,
    MatToolbarModule,
    MatProgressSpinnerModule
  ],
  providers: [
      { provide: APP_CONFIG, useValue: AppConfig },
      ReportService,
      DatasetService,
      TemplateService,
      ReportGenerationService
  ],
  bootstrap: [AppComponent]
})
export class AppModule { }
