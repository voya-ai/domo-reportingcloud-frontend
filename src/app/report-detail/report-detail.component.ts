import { Component, OnInit, EventEmitter } from '@angular/core';
import { Observable } from 'rxjs/Observable';
import { Subject } from 'rxjs/Subject';

import { ReportService } from 'app/report.service';

import { Report } from 'app/report';

import { Router, ActivatedRoute, ParamMap } from '@angular/router'

import 'rxjs/add/observable/fromPromise';
import 'rxjs/add/observable/fromEvent';
import 'rxjs/add/operator/switchMap';
import 'rxjs/add/operator/merge';

interface Timespan {
  value: number,
  label: string
}

@Component({
  selector: 'app-report-detail',
  templateUrl: './report-detail.component.html',
  styleUrls: ['./report-detail.component.css']
})
export class ReportDetailComponent implements OnInit {
  report$: Observable<Report>;
  reportEditSubject: Subject<Report> = new Subject<Report>();
  childGroupColumns: any[];

  ranges: Timespan[] = [{
    value: 7,
    label: "Weekly"
  },
  {
    value: 28,
    label: "4 weeks"
  },
  {
    value: 0,
    label: "All time"
  }];
  timespan: any = 0;
  startDate: string;
  timeColumn: string;
  masterGroup: string;

  loading: boolean = false;

  constructor(private _reportService: ReportService,
    private _route: ActivatedRoute,
    private _router: Router) {}

  ngOnInit() {
    this.report$ = this._route.paramMap
      .switchMap((params: ParamMap) =>
        this._reportService.getOne(params.get('slug')))
      .merge(this.reportEditSubject);

    this.report$
      .subscribe((report : Report) => {
        this.childGroupColumns = report.headerColumns.map((column) => {
          return { name: column, childGroup: false }
        })
      });
  }

  generate(report: Report) {
    this.loading = true;

    this._reportService.generate(report, {
      childGroups: this.childGroupColumns.filter(column => column.childGroup).map(column => column.name),
      timeColumn: this.timeColumn,
      startDate: this.startDate,
      masterGroup: this.masterGroup,
      timespan: this.timespan
    }).then(function(url) {
      this.loading = false;
      this._router.navigateByUrl('/');
    }.bind(this), function(error) {
      this.loading = false;
      this._router.navigateByUrl('/');
    });
  }

  update(report: Report) {
    this.reportEditSubject.next(report);
  }
}
