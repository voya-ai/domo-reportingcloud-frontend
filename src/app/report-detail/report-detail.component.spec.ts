import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { ReportDetailComponent } from './report-detail.component';
import { Observable } from 'rxjs/Observable';
import {
  MatSelectModule,
  MatDatepickerModule,
  MatCheckboxModule
} from '@angular/material';
import { ActivatedRoute, Router } from '@angular/router'

import { ActivatedRouteStub } from 'src/testing/router-stubs';
import { MockComponent } from 'src/testing/component-stubs';

import { Report } from 'app/report';
import { ReportService } from 'app/report.service';

import 'rxjs/add/observable/of';

class MockReportService {
  getOne(slug : string) : Observable<Report> {
    return Observable.of<Report>();
  }
}

class RouterStub {
  navigateByUrl(url: string) { return url; }
}

describe('ReportDetailComponent', () => {
  let component: ReportDetailComponent;
  let fixture: ComponentFixture<ReportDetailComponent>;

  beforeEach(async(() => {
    let activatedRoute = new ActivatedRouteStub();
    activatedRoute.testParamMap = { slug: 'invoice' };

    TestBed.configureTestingModule({
      providers: [
        { provide: ReportService, useClass: MockReportService },
        { provide: Router, useClass: RouterStub },
        { provide: ActivatedRoute, useValue: activatedRoute },
      ],
      declarations: [
        ReportDetailComponent,
        MockComponent({ selector: 'app-report-detail-sample', inputs: [ 'report' ] }),
        MockComponent({ selector: 'app-report-edit', inputs: [ 'report' ] })
      ],
      imports: [
        MatSelectModule,
        MatDatepickerModule,
        MatCheckboxModule,
        FormsModule,
        ReactiveFormsModule
      ]
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed
      .createComponent(ReportDetailComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match the snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
