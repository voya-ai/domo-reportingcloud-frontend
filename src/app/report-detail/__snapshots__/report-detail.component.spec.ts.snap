// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportDetailComponent should match the snapshot 1`] = `
<app-report-detail
  _reportService={[Function MockReportService]}
  _route={[Function ActivatedRouteStub]}
  _router={[Function RouterStub]}
  loading="false"
  ranges={[Function Array]}
  report$={[Function Observable]}
  reportEditSubject={[Function Subject]}
  timespan="0"
>
  
  
</app-report-detail>
`;
